File: app/src/main/java/com/balch/auctionbrowser/AuctionViewModel.java
package com.balch.auctionbrowser;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.VisibleForTesting;

import com.balch.auctionbrowser.auction.AuctionAdapter;
import com.balch.auctionbrowser.auction.model.EBayModel;
import com.balch.auctionbrowser.note.Note;
import com.balch.auctionbrowser.note.NotesModel;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

public class AuctionViewModel extends ViewModel {

    private static final int AUCTION_FETCH_COUNT = 30;

    private EBayModel auctionModel;
    private NotesModel notesModel;

    private AuctionAdapter auctionAdapter;

    private long totalPages;
    private int currentPage;
    private String searchText;
    private EBayModel.SortColumn sortColumn;

    private Disposable disposableGetAuction;

    private final MutableLiveData<AuctionData> auctionDataLive = new MutableLiveData<>();

    boolean isInitialized() {
        return ((auctionModel != null) && (notesModel != null));
    }

    public void setAuctionAdapter(AuctionAdapter auctionAdapter) {
        this.auctionAdapter = auctionAdapter;
    }

    public AuctionAdapter getAuctionAdapter() {
        return auctionAdapter;
    }

    void setAuctionModel(EBayModel auctionModel) {
        this.auctionModel = auctionModel;
    }

    void setNotesModel(NotesModel notesModel) {
        this.notesModel = notesModel;
    }

    LiveData<AuctionData> getAuctionData() {
        return auctionDataLive;
    }

    void loadAuctions(EBayModel.SortColumn sortColumn) {
        loadAuctions(searchText, sortColumn);
    }

    void loadAuctions(String searchText, EBayModel.SortColumn sortColumn) {
        this.totalPages = -1;
        this.currentPage = 1;
        this.searchText = searchText;
        this.sortColumn = sortColumn;
        getAuctionsAsync();
    }

    void loadAuctionsNextPage() {
        this.currentPage++;
        getAuctionsAsync();
    }

    String getSearchText() {
        return searchText;
    }

    boolean hasMoreAuctionPages(long page) {
        return ((totalPages == -1) || (page < totalPages));
    }

    void insertNote(Note note) {
        notesModel.insert(note);
    }

    void updateNote(Note note) {
        notesModel.update(note);
    }

    void deleteNote(Note note) {
        notesModel.delete(note);
    }

    private void getAuctionsAsync() {
        disposeGetAuctionDisposable();
        disposableGetAuction = auctionModel
                .getAuctions(searchText, currentPage, AUCTION_FETCH_COUNT, sortColumn)
                .subscribeOn(Schedulers.io())
                .map(new Function<AuctionData, AuctionData>() {
                    @Override
                    public AuctionData apply(@NonNull AuctionData auctionData) throws Exception {
                        if (auctionData.getAuctions() != null) {
                            auctionData.setNotes(notesModel.getNotes(auctionData.getAuctions()));
                        }
                        totalPages = auctionData.getTotalPages();

                        return auctionData;
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<AuctionData>() {
                               @Override
                               public void accept(@NonNull AuctionData auctionData) throws Exception {
                                   auctionDataLive.setValue(auctionData);
                               }
                           },
                            new Consumer<Throwable>() {
                                @Override
                                public void accept(@NonNull Throwable throwable) throws Exception {
                                    auctionDataLive.setValue(null);
                                }
                            });
    }

    void disposeGetAuctionDisposable() {
        if (disposableGetAuction != null) {
            disposableGetAuction.dispose();
            disposableGetAuction = null;
        }
    }

    @Override
    protected void onCleared() {
        disposeGetAuctionDisposable();
    }

    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
    public NotesModel getNotesModel() {
        return notesModel;
    }
}

File: app/src/main/java/com/balch/auctionbrowser/AuctionViewModel.kt
/*
 * Author: Balch
 *
 * This file is part of AuctionBrowser.
 *
 * AuctionBrowser is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AuctionBrowser is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MockTrade.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2017
 *
 */

package com.balch.auctionbrowser

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.support.annotation.VisibleForTesting
import android.util.Log
import com.balch.auctionbrowser.auction.AuctionAdapter
import com.balch.auctionbrowser.auction.model.EBayModel
import com.balch.auctionbrowser.note.Note
import com.balch.auctionbrowser.note.NotesModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

/**
 * This ViewModel exposes a LiveData object which emits AuctionData objects from the
 * EBay API. An AuctionData object contains the current page of found Auctions.
 *
 * The ViewModel also stores state data that should survive a ConfigChange operation. This
 * includes the AuctionAdapter that contains the entire Auction list and the info necessary
 * to retrieve the next page of Auctions.
 *
 * The class implements simple dependency injection using the `inject()` method to setter-inject
 * the Adapter and ModelApis.
 */
class AuctionViewModel : ViewModel() {

    private val AUCTION_FETCH_COUNT = 30

    // public properties
    var isInitialized = false
        private set

    val auctionData: LiveData<AuctionData>
        get() = auctionDataLive

    var searchText: String? = null
        private set

    // injected models
    lateinit private var auctionModel: EBayModel
    @get:VisibleForTesting(otherwise = VisibleForTesting.NONE)
    lateinit var notesModel: NotesModel
        private set
    lateinit var auctionAdapter: AuctionAdapter
        private set

    // paging vars
    private var totalPages: Long = 0
    private var currentPage: Int = 0
    private var sortColumn: EBayModel.SortColumn = EBayModel.SortColumn.BEST_MATCH

    // LiveData<AuctionData>
    private val auctionDataLive = MutableLiveData<AuctionData>()

    // disposables
    private var disposableGetAuction: Disposable? = null

    fun inject(adapter: AuctionAdapter, eBayModel: EBayModel, notesModel: NotesModel) {
        isInitialized = true
        this.auctionAdapter = adapter
        this.auctionModel = eBayModel
        this.notesModel = notesModel
    }

    fun loadAuctions(searchText: String? = null, sortColumn: EBayModel.SortColumn? = null) {
        this.totalPages = -1
        this.currentPage = 1
        this.searchText = searchText ?: this.searchText ?: ""
        this.sortColumn = sortColumn ?: this.sortColumn
        getAuctionsAsync()
    }

    fun loadAuctionsNextPage() {
        this.currentPage++
        getAuctionsAsync()
    }

    fun hasMoreAuctionPages(page: Long): Boolean {
        return totalPages == -1L || page < totalPages
    }

    fun insertNote(note: Note) {
        notesModel.insert(note)
    }

    fun updateNote(note: Note) {
        notesModel.update(note)
    }

    fun deleteNote(note: Note) {
        notesModel.delete(note)
    }

    private fun getAuctionsAsync() {
        disposeGetAuctionDisposable()
        disposableGetAuction = auctionModel
                .getAuctions(searchText!!, currentPage.toLong(), AUCTION_FETCH_COUNT, sortColumn)
                .subscribeOn(Schedulers.io())
                .doOnNext { auctionData ->
                    auctionData.notes = notesModel.getNotes(auctionData.auctions)
                    totalPages = auctionData.totalPages.toLong()
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({auctionData -> auctionDataLive.setValue(auctionData)},
                            { throwable ->
                                Log.e("AuctionViewModel", "Error", throwable)
                                auctionDataLive.setValue(null)
                            })
    }

    fun disposeGetAuctionDisposable() {
        disposableGetAuction?.dispose()
        disposableGetAuction = null
    }

    override fun onCleared() {
        disposeGetAuctionDisposable()
    }

}


